[Flow] Add module.http component (#4297)
grafana/agent: docs/sources/flow/reference/components/module.http.md

---
title: module.http
labels:
  stage: beta
---

# module.http

{{< docs/shared lookup="flow/stability/beta.md" source="agent" >}}

> `module.http` is a *module loader* component. A module loader is a Grafana Agent Flow
> component which retrieves a [module][] and runs the components defined inside of it.

Reviewer: ```suggestion
`module.http` is a *module loader* component available with Grafana Agent in flow mode. 
The loader retrieves a [module][] and runs the components defined in the module.
```

Could we rephrase this a little? Does this lose the meaning and intent of the definition?

Author: The second sentence sounds good to me :smile:. For the first, I realized now there is a [Module Loader section in our existing docs](https://grafana.com/docs/agent/latest/flow/concepts/modules/#module-loaders) that I think makes more sense to just link to, rather than re-explain.

Reviewer: I agree. Linking to the Module Loader section is better. I'll get to editing that existing doc at some point as well :-)

> `module.http` embeds a [remote.http][] component to retrieve the module from a remote
> HTTP server. This allows for the use of a single module loader, rather than a `remote.http`
> component paired with a [module.string][] component.

Reviewer: ```suggestion
`module.http` embeds a [remote.http][] component to retrieve the module from a remote
HTTP server. This module allows you to use a single module loader, rather than a `remote.http`
component paired with a [module.string][] component.
```

`url` | `string` | URL to poll. | | yes
`method` | `string` | Define HTTP method for the request | `"GET"` | no
`headers` | `map(string)` | Custom headers for the request. | `{}` | no
`poll_frequency` | `duration` | Frequency to poll the URL. | `"1m"` | no
`poll_timeout` | `duration` | Timeout when polling the URL. | `"10s"` | no
`is_secret` | `bool` | Whether the response body should be treated as a secret. | false | no

[secret]: {{< relref "../../config-language/expressions/types_and_values.md#secrets" >}}

## Blocks

The following blocks are supported inside the definition of `module.http`:

Hierarchy        | Block      | Description | Required
---------------- | ---------- | ----------- | --------
arguments | [arguments][] | Arguments to pass to the module. | no

[arguments]: #arguments-block

### arguments block

The `arguments` block specifies the list of values to pass to the loaded
module.

The attributes provided in the `arguments` block are validated based on the
[argument blocks][] defined in the module source:

* If a module source marks one of its arguments as required, it must be
  provided as an attribute in the `arguments` block of the module loader.

> * Attributes in the `argument` block of the module loader will be rejected if

Reviewer: ```suggestion
* Attributes in the `argument` block of the module loader are rejected if
```

  they are not defined in the module source.

[argument blocks]: {{< relref "../config-blocks/argument.md" >}}

## Exported fields

The following fields are exported and can be referenced by other components:

Name | Type | Description
---- | ---- | -----------
`exports` | `map(any)` | The exports of the Module loader.

`exports` exposes the `export` config block inside a module. It can be accessed
from the parent config via `module.http.LABEL.exports.EXPORT_LABEL`.

Values in `exports` correspond to [export blocks][] defined in the module
source.

[export blocks]: {{< relref "../config-blocks/export.md" >}}

## Component health

`module.http` is reported as healthy if the most recent load of the module was
successful.

Before the first load of the module, the health is reported as `Unknown`.

If the module is not loaded successfully, the current health displays as

> unhealthy and the health includes the error from loading the module.

Reviewer: ```suggestion
unhealthy, and the health includes the error from loading the module.
```
